# QC report
- study: `r study`
- author: `r author`
- date: `r format(Sys.time(), '%d %B, %Y')`

## Parameters used for QC

```{r, results="markup"}
qc.summary$parameters
```


## Sex mismatches

There are `r sum(qc.summary$sex.summary$tab$outliers)` sex detection outliers, and `r sum(qc.summary$sex.summary$tab$sex.mismatch == "TRUE")` sex detection mismatches.

```{r}

tab <- qc.summary$sex.check
if(nrow(tab) > 0) kable(tab,row.names=F)

```

This is a plot of the difference between median 
chromosome Y and chromosome X probe intensities ("XY diff").
Cutoff for sex detection was
XY diff = `r qc.summary$parameters$sex.cutoff`.

```{r,message=FALSE}

(qc.summary$sex.summary$graph)

```


## Methylated vs unmethylated

There are `r sum(qc.summary$meth.unmeth.summary$tab$outliers)` outliers from the meth vs unmeth comparison.
Outliers are samples whose predicted median methylated signal is
more than `r qc.summary$parameters$meth.unmeth.outlier.sd` standard deviations
from the expected (regression line).

```{r}

tab <- subset(qc.summary$meth.unmeth.summary$tab, outliers)
if(nrow(tab) > 0) kable(tab,row.names=F)

```

This is a plot of the methylation signals vs unmethylated signals

```{r,message=FALSE}

(qc.summary$meth.unmeth.summary$graph)

```


## Control probe means

There were `r sum(qc.summary$controlmeans.summary$tab$outliers)` outliers detected based on deviations from mean values for control probes.

```{r}

tab <- subset(qc.summary$controlmeans.summary$tab, outliers)
if(nrow(tab) > 0) kable(tab,row.names=F)

```

The distribution of sample control means are plotted here:

```{r, message=FALSE, fig.height=24}

(qc.summary$controlmeans.summary$graph)

```


## Sample detection p-values

There were `r sum(qc.summary$sample.detectionp.summary$tab$outliers)` samples
with a high proportion of undetected probes
(proportion of probes with
detection p-value > `r qc.summary$parameters$detection.threshold`
is > `r qc.summary$parameters$detectionp.samples.threshold`).

```{r}

tab <- subset(qc.summary$sample.detectionp.summary$tab, outliers)
if(nrow(tab) > 0) kable(tab,row.names=F)

```

Distribution:

```{r, message=FALSE}

(qc.summary$sample.detectionp.summary$graph)

```


## Sample bead numbers

There were `r sum(qc.summary$sample.beadnum.summary$tab$outliers)` samples
with a high proportion of probes with low bead number
(proportion of probes with
bead number < `r qc.summary$parameters$bead.threshold`
is > `r qc.summary$parameters$beadnum.samples.threshold`).

```{r}

tab <- subset(qc.summary$sample.beadnum.summary$tab, outliers)
if(nrow(tab) > 0) kable(tab,row.names=F)

```

Distribution:

```{r, message=FALSE}

(qc.summary$sample.beadnum.summary$graph)

```


## CpG detection p-values

There were `r sum(qc.summary$cpg.detectionp.summary$tab$outliers)`
probes with only background signal in a high proportion of samples
(proportion of samples with detection p-value > `r qc.summary$parameters$detection.threshold`
is > `r qc.summary$parameters$detectionp.cpgs.threshold`).
Manhattan plot shows the proportion of samples.

```{r,message=FALSE}

if (!is.null(qc.summary$cpg.detectionp.summary$graph))
    (qc.summary$cpg.detectionp.summary$graph)

```

## Low number of beads per CpG

There were `r sum(qc.summary$cpg.beadnum.summary$tab$outliers)` CpGs
with low bead numbers in a high proportion of samples
(proportion of samples with bead number < `r qc.summary$parameters$bead.threshold`
is > `r qc.summary$parameters$beadnum.cpgs.threshold`).
Manhattan plot of proportion of samples.

```{r,message=FALSE}

if (!is.null(qc.summary$cpg.beadnum.summary$graph))
    (qc.summary$cpg.beadnum.summary$graph)

```

## Cell count estimates

```{r,message=FALSE}
child.filename <- file.path(report.path, "missing.rmd")
if (!is.null(qc.summary$cell.counts))
    child.filename <- file.path(report.path, "cell-counts.rmd")
```

```{r, child = child.filename}
``` 

## SNP probe beta values

Distributions of SNP probe beta values.
```{r,message=FALSE}

(qc.summary$genotype.summary$graphs$snp.beta)

```

## Genotype concordance

```{r,message=FALSE}
child.filename <- file.path(report.path, "missing.rmd")
if (!is.null(qc.summary$genotype.summary$graphs$snp.concordance))
    child.filename <- file.path(report.path, "genotype-concordance.rmd")
```

```{r, child=child.filename}
```

## R session information

```{r, results="markup"}
sessionInfo()
```
